src/omv/sensor.c

```diff
@@ -583,26 +599,30 @@ int sensor_set_framesize(framesize_t framesize)
     // Skip the first frame.
     MAIN_FB()->bpp = -1;
 
-    // Set MAIN FB x, y offset.
+    // Set MAIN FB x offset, y offset, width, height, backup width, and backup height.
     MAIN_FB()->x = 0;
     MAIN_FB()->y = 0;
+    MAIN_FB()->w = MAIN_FB()->u = resolution[framesize][0];
+    MAIN_FB()->h = MAIN_FB()->v = resolution[framesize][1];
 
-    // Set MAIN FB width and height.
-    MAIN_FB()->w = resolution[framesize][0];
-    MAIN_FB()->h = resolution[framesize][1];
-
-    // Set MAIN FB backup width and height.
-    MAIN_FB()->u = resolution[framesize][0];
-    MAIN_FB()->v = resolution[framesize][1];
     return 0;
 }
 
 int sensor_set_windowing(int x, int y, int w, int h)
 {
-    MAIN_FB()->x = x;
-    MAIN_FB()->y = y;
-    MAIN_FB()->w = MAIN_FB()->u = w;
-    MAIN_FB()->h = MAIN_FB()->v = h;
+    // py_sensor_set_windowing ensures this the window is at least 8x8
+    // and that it is fully inside the sensor output framesize window.
+    if (sensor.pixformat == PIXFORMAT_JPEG) {
+        return -1;
+    }
+
+    // We force everything to be a multiple of 2 so that when you switch between
+    // grayscale/rgb565/bayer/jpeg the frame doesn't need to move around for bayer to work.
+    MAIN_FB()->x = (x / 2) * 2;
+    MAIN_FB()->y = (y / 2) * 2;
+    MAIN_FB()->w = MAIN_FB()->u = (w / 2) * 2;
+    MAIN_FB()->h = MAIN_FB()->v = (h / 2) * 2;
+
```

Comment:
I don't understand, there's not need to skip a frame here.

Note that using the cropped flag would seem nice... but, then it has to be maintained by another method. I could add a method that returns if the image is being cropped. Will do that.

Reply:
Added a method. The /2 part needs to be kept as sensor_check_buffsize() doesn't run unless the image is too big. The point of the /2 stuff is to prevent BAYER versus GRAYSCALE from having different scaling/cropping values. It looks bad when the image changes size/shape/moves when you change the pixformat. With this code everything looks good and makes sense to the user.